### DISCUSSION TITLE
Arduino Due's native port reports full speed, but 512 byte max packet size

### DISCUSSION THREAD

**Discussion by phatchman** [hathach/tinyusb#1900] [General]
### Operating System

Linux

### Board

Pico

### Firmware

examples/hosts/cdc_msc_hid modified to receive 512 + 1 bytes. (originally receives 64 + 1 bytes)

### What happened ?

This is not a bug in tiny usb, but rather the Arduino Due. It reports as a Full Speed device, but reports a max packet size of 512. It also sends with a 512 byte packet size. The following patch allows tinyusb to communicate with the due in host mode. It modifies the max packet size from the descriptor from 512 bytes to 64 bytes.

It is also necessary to call tuh_cdc_read with a buffer size of at least 512 bytes (rather than 64) to ensure that data is not lost.

```
diff --git a/src/class/cdc/cdc_host.c b/src/class/cdc/cdc_host.c
index e9c3d34cb..84d2a4fcc 100644
--- a/src/class/cdc/cdc_host.c
+++ b/src/class/cdc/cdc_host.c
@@ -503,7 +503,14 @@ bool cdch_open(uint8_t rhport, uint8_t daddr, tusb_desc_interface_t const *itf_d
       tusb_desc_endpoint_t const *desc_ep = (tusb_desc_endpoint_t const *) p_desc;
       TU_ASSERT(TUSB_DESC_ENDPOINT == desc_ep->bDescriptorType &&
                 TUSB_XFER_BULK     == desc_ep->bmAttributes.xfer);
-
+      
+      // The Arduino Due reports Full speed, and packet size of 512 byes, which is invalid.
+      if ( TUSB_SPEED_FULL == tuh_speed_get(daddr) && desc_ep->wMaxPacketSize > 64 )
+      {
+        // Modify the enpoint descriptor to set max packet size to 64 bytes
+        tusb_desc_endpoint_t *tmp_desc_ep = (tusb_desc_endpoint_t *) p_desc;
+        tmp_desc_ep->wMaxPacketSize = 64;
+      }
       TU_ASSERT(tuh_edpt_open(daddr, desc_ep));
 
       if ( tu_edpt_dir(desc_ep->bEndpointAddress) == TUSB_DIR_IN )

```

### How to reproduce ?

Connect to native port on the Arduino Due.
Run modified cdc host example.

### Debug Log as txt file

[tinyusb_due.log](https://github.com/hathach/tinyusb/files/10713143/tinyusb_due.log)

### I have checked existing issues, dicussion and documentation

- [X] I confirm I have checked existing issues, dicussion and documentation.

**Comment by hathach** [comment]
max FS bulk endpoint size is only 64 bytes. This violate the USB specs, 2 wrongs cannot make a right. You should make an PR/issue to Arduino Due for them to fix the issue. The suggested patch is not accepted.

**Reply by phatchman** [reply]
OK. I understand and agree.
